### PR TITLE
add teardown/0 to release.ex

### PR DESCRIPTION
**Pull request by jinjagit**
This addresses issue [#88](https://github.com/liquidvotingio/api/issues/88)

### CODE REVIEW

**Comment by jinjagit** [comment]
Would we prefer this def to have organization_id accepted as arg? e.g. call teardown("bc7eeccb-5e10-4004-8bfb-7fc68536bbd7")

File: lib/liquid_voting/release.ex [hunk 1]
```diff
@@ -11,6 +13,23 @@ defmodule LiquidVoting.Release do
...
+  For use before and after running smoke tests.
+  """
+  def teardown() do
+    organization_id = "bc7eeccb-5e10-4004-8bfb-7fc68536bbd7"
+
+    votes = Voting.list_votes(organization_id)
+    Enum.each(votes, fn vote -> Voting.delete_vote!(vote) end)
+
+    participants = Voting.list_participants(organization_id)
+    Enum.each(participants, fn participant -> Voting.delete_participant(participant) end)
```
Comment (niklaslong) [inline lib/liquid_voting/release.ex:27]: We should probably use the `!` version of this (or pattern match on the response at the very least)?

Comment (niklaslong) [inline lib/liquid_voting/release.ex:27]: We should probably use the `!` version of this (or pattern match on the response at the very least)?

Comment (jinjagit) [reply lib/liquid_voting/release.ex:27]: I guess pattern match then, since there is no ! version

**Comment by niklaslong** [comment]
I'm in favour of passing in the `organisation_id`; that way we're not dependant on a hardcoded value.

File: lib/liquid_voting/release.ex (hunk 1 above)
Comment (jinjagit) [reply lib/liquid_voting/release.ex]: Not sure I unsderstand the use of pattern matching the response, since it will be a list (empty, or with contents).

Comment (jinjagit) [reply lib/liquid_voting/release.ex]: Added pattern matching for return value of calls to delete_participant/1, and organization_id is now passed in to the def.

**Comment by oliverbarnes** [comment]
The benefit of sticking to a hardcoded value would be that teardown can only be called on that given org, even if a malicious user somehow gains access to the release app (or to the smoke tests themselves). 

Then again, if someone gains access to the release they're also able to run `rollback` 😅, so that scenario may be moot. 

Mistakes on the smoke test scripts may still be an issue? Like, pasting the wrong org id by mistake and deleting a whole real organization's data in the process.... For that, the developer would need to have an organization id in their clipboard. From, say, browsing the production auth database.

Perhaps both the demo org id and test org ids should be permanent and hard coded, is what I'm saying. And better not be flexible with them. What do you think?

**Comment by jinjagit** [comment]
I agree. Hardcoding is best , to prevent accidental deletion (I went on the same mental journey as you, Oliver, re. malicious use ;-) )

File: lib/liquid_voting/release.ex (hunk 1 above)
Comment (niklaslong) [reply lib/liquid_voting/release.ex]: It's to throw an error if something goes wrong, otherwise we would have no way of knowing.

Comment (jinjagit) [reply lib/liquid_voting/release.ex]: Yeah, eventually I worked it out, I think.

**Comment by niklaslong** [comment]
True, those reasons definitely outweigh the benefits of not having it hardcoded. 👍

**Comment by oliverbarnes** [comment]
Cool 👍 

> I went on the same mental detour as you, Oliver, re. malicious use ;-)

malicious use speculation always fun (and labyrinthic) 

> I'll hardcode the test and demo ids, then we can pass a value in to switch to either - or some similar idea.

Actually, I don't mean hardcoding both demo and test orgs id here - we just need the test org id in the teardown function. I mean the demo org id should likewise be permanent and hardcoded where relevant (docs).

**Comment by jinjagit** [comment]
organization_id value now hard-coded into teardown/0

File: lib/liquid_voting/release.ex [hunk 2]
```diff
@@ -11,6 +13,26 @@ defmodule LiquidVoting.Release do
     {:ok, _, _} = Ecto.Migrator.with_repo(repo, &Ecto.Migrator.run(&1, :down, to: version))
   end
 
+  @doc """
+  Deletes votes, participants and delegations for the smoke tests organization_id.
+  For use before and after running smoke tests.
+  """
+  def teardown() do
+    organization_id = "bc7eeccb-5e10-4004-8bfb-7fc68536bbd7"
```
Comment (oliverbarnes) [inline lib/liquid_voting/release.ex:21]: it's good practice to define constants outside of functions, to keep them nice and visible when browsing the source. Here you could extract this and define it as a module attribute like this:

```
defmodule LiquidVoting.Release do
  @app :liquid_voting
  @test_organization_id "bc7eeccb-5e10-4004-8bfb-7fc68536bbd7"
```

(prepending _test__ makes it more explicit too)

File: lib/liquid_voting/release.ex [hunk 3]
```diff
@@ -11,6 +13,26 @@ defmodule LiquidVoting.Release do
     {:ok, _, _} = Ecto.Migrator.with_repo(repo, &Ecto.Migrator.run(&1, :down, to: version))
   end
 
+  @doc """
+  Deletes votes, participants and delegations for the smoke tests organization_id.
+  For use before and after running smoke tests.
+  """
+  def teardown() do
```
Comment (oliverbarnes) [inline lib/liquid_voting/release.ex:20]: wondering if `smoke_test_teardown()` would be clearer

**Comment by jinjagit** [comment]
I think this ready to merge now.

File: lib/liquid_voting/release.ex [hunk 4]
```diff
@@ -11,6 +13,26 @@ defmodule LiquidVoting.Release do
...
+    organization_id = "bc7eeccb-5e10-4004-8bfb-7fc68536bbd7"
+
+    votes = Voting.list_votes(organization_id)
+    Enum.each(votes, fn vote -> Voting.delete_vote!(vote) end)
```
Comment (oliverbarnes) [inline lib/liquid_voting/release.ex:24]: Maybe 

```
organization_id
|> Voting.list_votes
|> Enum.each(fn vote -> Voting.delete_vote!(vote) end)
```

@niklaslong what do you think?

Comment (oliverbarnes) [reply lib/liquid_voting/release.ex:24]: and if so, same for participants and delegations

Comment (jinjagit) [reply lib/liquid_voting/release.ex:23]: function renamed, and organization_id set as module attribute

Comment (jinjagit) [reply lib/liquid_voting/release.ex]: refactored to use pipe construction

**Comment by jinjagit** [comment]
FYI... the mix task equivalent included `import Ecto.Query, warn: false`, and also ran the app with: `Mix.Task.run("app.start", [])`

**Comment by oliverbarnes** [comment]
> FYI... the mix task equivalent includes `import Ecto.Query, warn: false`, and also runs the app with: `Mix.Task.run("app.start", [])`

mix & releases are different beasts. it's a good point to get familiar with them:

https://hexdocs.pm/phoenix/releases.html
https://hexdocs.pm/mix/Mix.Tasks.Release.html

**Comment by jinjagit** [comment]
Yes, i mean maybe we need to start the app, etc. in the release code. E.g.
```
# To start your system
    _build/prod/rel/my_app/bin/my_app start
```

**Comment by niklaslong** [comment]
I think we'll need something similar to what is implemented for `Ecto.Migrator.with_repo/3`, i.e. ensuring repo is started and if not start manually. 

https://github.com/elixir-ecto/ecto_sql/blob/2e1378eb5d65c2f34d49b595891e7ab482b44507/lib/ecto/migrator.ex#L124

**Comment by jinjagit** [comment]
This appears to be doing something now.
`eval "LiquidVoting.Release.smoke_test_teardown"` now returns `12:38:29.673 [info] Already up`

Since I have only a vague idea of what I am really doing, here is what I did / how I tested it:

1. edited release.ex using Niklas's observation, in previous comment.
2. started production locally using Oliver's post with relevant commands
3. opened liquid_voting_dev on psql and checked participants table is empty
4. started api server locally
5. ran k6 delegations_test
6. checked participants table now included new participants (true)
7. ran the eval command (and preceding code block), mentioned at top of this comment.
8. checked participants table now empty (true)

**Comment by niklaslong** [comment]
It does seem like progress! 💪 
I don't think we actually need to run a migration though. How about:
```elixir
def smoke_test_teardown() do
  for repo <- repos() do
    {:ok, _, _} = Ecto.Migrator.with_repo(repo, fn _repo -> run_teardown() end)
  end
end
```

**Comment by jinjagit** [comment]
> It does seem like progress! muscle
> I don't think we actually need to run a migration though. How about:
> 
> ```elixir
> def smoke_test_teardown() do
>   for repo <- repos() do
>     {:ok, _, _} = Ecto.Migrator.with_repo(repo, fn _repo -> run_teardown() end)
>   end
> end
> ```

Nice. I was wondering about that. Seems to work fine without migration.

File: lib/liquid_voting/release.ex [hunk 5]
```diff
@@ -11,6 +14,32 @@ defmodule LiquidVoting.Release do
     {:ok, _, _} = Ecto.Migrator.with_repo(repo, &Ecto.Migrator.run(&1, :down, to: version))
   end
 
+  @doc """
+  Deletes votes, participants and delegations for the smoke tests organization_id.
+  For use before and after running smoke tests.
+  """
+  def smoke_test_teardown() do
+    for repo <- repos() do
+      {:ok, _, _} = Ecto.Migrator.with_repo(repo, fn _repo -> run_teardown() end)
```
Comment (oliverbarnes) [inline lib/liquid_voting/release.ex:23]: 👏

File: lib/liquid_voting/release.ex [hunk 6]
```diff
@@ -11,6 +14,32 @@ defmodule LiquidVoting.Release do
...
+    end
+  end
+
+  defp run_teardown() do
+    @test_organization_id
```
Comment (oliverbarnes) [inline lib/liquid_voting/release.ex:28]: It'd be good to have some output to sure we know this ran sucessfully. Maybe a basic count of resources before the teardown runs, and another after? Output could be:

```
## About to run smoke test teardown 
Current resource counts: 
10 Participants, 5 Votes, 3 Delegations
## Running smoke test teardown on test organization data
## Teardown ran successfully.
End resource counts: 
0 Participants, 0 Votes, 0 Delegations
## Have a nice day
```

Comment (jinjagit) [reply lib/liquid_voting/release.ex:28]: sure.

Comment (jinjagit) [reply lib/liquid_voting/release.ex:57]: After running all 4 smoke tests, `eval "LiquidVoting.Release.smoke_test_teardown"` now outputs: 
teardown-message

**Comment by jinjagit** [comment]
Is the organization of code / defs OK?

**Comment by oliverbarnes** [comment]
> Is the organization of code / defs OK?

Yeah the printing of resources being a separate function separates concerns just enough, keeping things DRY. It does run the same queries for listing resources 3 times instead of twice (the initial listing could be re-used for the deletion after the count printout), but I wasn't too concerned about that - these will be very quick queries on a tiny number of rows each time.

**Comment by jinjagit** [comment]
Cool. Thanks for checking.

File: lib/liquid_voting/release.ex (hunk 6 above)
Comment (oliverbarnes) [reply lib/liquid_voting/release.ex:57]: 🍭 sweet

Comment (niklaslong) [reply lib/liquid_voting/release.ex:57]: Love the polite code ❤️

**Pull request merged**